<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: An R package for gene and isoform differential expression
analysis of RNA-seq data</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> An R package for gene and isoform differential expression
analysis of RNA-seq data
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;EBSeq&rsquo; version 1.14.0</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../doc/index.html">User guides, package vignettes and other documentation.</a></li>
<li><a href="../demo">Code demos</a>.  Use <a href="../../utils/help/demo">demo()</a> to run them.</li>
<li><a href="../NEWS">Package NEWS</a>.</li>
</ul>

<h2>Help Pages</h2>


<table width="100%">
<tr><td style="width: 25%;"><a href="EBSeq_NingLeng-package.html">EBSeq_NingLeng-package</a></td>
<td>EBSeq: RNA-Seq Differential Expression Analysis on both gene and isoform level</td></tr>
<tr><td style="width: 25%;"><a href="beta.mom.html">beta.mom</a></td>
<td>Fit the beta distribution by method of moments</td></tr>
<tr><td style="width: 25%;"><a href="crit_fun.html">crit_fun</a></td>
<td>Calculate the soft threshold for a target FDR</td></tr>
<tr><td style="width: 25%;"><a href="DenNHist.html">DenNHist</a></td>
<td>Density plot to compare the empirical q's and the simulated q's from the fitted beta distribution.</td></tr>
<tr><td style="width: 25%;"><a href="EBMultiTest.html">EBMultiTest</a></td>
<td>Using EM algorithm to calculate the posterior probabilities of interested patterns in a multiple condition study</td></tr>
<tr><td style="width: 25%;"><a href="EBSeq_NingLeng-package.html">EBSeq_NingLeng</a></td>
<td>EBSeq: RNA-Seq Differential Expression Analysis on both gene and isoform level</td></tr>
<tr><td style="width: 25%;"><a href="EBTest.html">EBTest</a></td>
<td>Using EM algorithm to calculate the posterior probabilities of being DE</td></tr>
<tr><td style="width: 25%;"><a href="f0.html">f0</a></td>
<td>The Prior Predictive Distribution of being EE</td></tr>
<tr><td style="width: 25%;"><a href="f1.html">f1</a></td>
<td>The Prior Predictive Distribution of being DE</td></tr>
<tr><td style="width: 25%;"><a href="GeneMat.html">GeneMat</a></td>
<td>The simulated data for two condition gene DE analysis</td></tr>
<tr><td style="width: 25%;"><a href="GetDEResults.html">GetDEResults</a></td>
<td>Obtain Differential Expression Analysis Results in a Two-condition Test</td></tr>
<tr><td style="width: 25%;"><a href="GetMultiFC.html">GetMultiFC</a></td>
<td>Calculate the Fold Changes for Multiple Conditions</td></tr>
<tr><td style="width: 25%;"><a href="GetMultiPP.html">GetMultiPP</a></td>
<td>Posterior Probability of Each Transcript</td></tr>
<tr><td style="width: 25%;"><a href="GetNg.html">GetNg</a></td>
<td>Ng Vector</td></tr>
<tr><td style="width: 25%;"><a href="GetNormalizedMat.html">GetNormalizedMat</a></td>
<td>Calculate normalized expression matrix</td></tr>
<tr><td style="width: 25%;"><a href="GetPatterns.html">GetPatterns</a></td>
<td>Generate all possible patterns in a multiple condition study</td></tr>
<tr><td style="width: 25%;"><a href="GetPP.html">GetPP</a></td>
<td>Generate the Posterior Probability of each transcript.</td></tr>
<tr><td style="width: 25%;"><a href="GetPPMat.html">GetPPMat</a></td>
<td>Posterior Probability of Transcripts</td></tr>
<tr><td style="width: 25%;"><a href="IsoList.html">IsoList</a></td>
<td>The simulated data for two condition isoform DE analysis</td></tr>
<tr><td style="width: 25%;"><a href="IsoMultiList.html">IsoMultiList</a></td>
<td>The simulated data for multiple condition isoform DE analysis</td></tr>
<tr><td style="width: 25%;"><a href="Likefun.html">Likefun</a></td>
<td>Likelihood Function of the NB-Beta Model</td></tr>
<tr><td style="width: 25%;"><a href="LikefunMulti.html">LikefunMulti</a></td>
<td>Likelihood Function of the NB-Beta Model In Multiple Condition Test</td></tr>
<tr><td style="width: 25%;"><a href="LogN.html">LogN</a></td>
<td>The function to run EM (one round) algorithm for the NB-beta model.</td></tr>
<tr><td style="width: 25%;"><a href="LogNMulti.html">LogNMulti</a></td>
<td>EM algorithm for the NB-beta model in the multiple condition test</td></tr>
<tr><td style="width: 25%;"><a href="MedianNorm.html">MedianNorm</a></td>
<td>Median Normalization</td></tr>
<tr><td style="width: 25%;"><a href="MultiGeneMat.html">MultiGeneMat</a></td>
<td>The simulated data for multiple condition gene DE analysis</td></tr>
<tr><td style="width: 25%;"><a href="PlotPattern.html">PlotPattern</a></td>
<td>Visualize the patterns</td></tr>
<tr><td style="width: 25%;"><a href="PlotPostVsRawFC.html">PlotPostVsRawFC</a></td>
<td>Plot Posterior FC vs FC</td></tr>
<tr><td style="width: 25%;"><a href="PolyFitPlot.html">PolyFitPlot</a></td>
<td>Fit the mean-var relationship using polynomial regression</td></tr>
<tr><td style="width: 25%;"><a href="PostFC.html">PostFC</a></td>
<td>Calculate the posterior fold change for each transcript across conditions</td></tr>
<tr><td style="width: 25%;"><a href="QQP.html">QQP</a></td>
<td>The Quantile-Quantile Plot to compare the empirical q's and simulated q's from fitted beta distribution</td></tr>
<tr><td style="width: 25%;"><a href="QuantileNorm.html">QuantileNorm</a></td>
<td>Quantile Normalization</td></tr>
<tr><td style="width: 25%;"><a href="RankNorm.html">RankNorm</a></td>
<td>Rank Normalization</td></tr>
</table>
</body></html>
